packages/runtime/runtime-utils/src/dataStoreRegistry.ts

```diff
@@ -0,0 +1,62 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import {
+    FluidDataStoreRegistryEntry,
+    IFluidDataStoreRegistry,
+    FluidDataStoreRegistry,
+    FluidDataStoreRegistryEntries,
+    IProvideFluidDataStoreRegistry,
+} from "@fluidframework/runtime-definitions";
+
+export function createDataStoreRegistry(arg: FluidDataStoreRegistry): IProvideFluidDataStoreRegistry {
+    // Ensure that FluidDataStoreRegistry type did not change from original
+    // We do casting below and need to ensure that only these two types can come in
+    const entries: FluidDataStoreRegistryEntries | IProvideFluidDataStoreRegistry = arg;
+
+    const registry = entries as IProvideFluidDataStoreRegistry;
+    if (registry.IFluidDataStoreRegistry !== undefined) {
+        return registry;
+    }
+    return new DataStoreRegistry(entries as FluidDataStoreRegistryEntries);
+}
+
+class DataStoreRegistry implements IFluidDataStoreRegistry {
+    private readonly map: Map<string, FluidDataStoreRegistryEntry | Promise<FluidDataStoreRegistryEntry>> = new Map();
+
+    public get IFluidDataStoreRegistry() { return this; }
+
+    constructor(namedEntries: FluidDataStoreRegistryEntries) {
+        for (const entry of namedEntries) {
+            if (Array.isArray(entry)) {
+                this.map.set(entry[0], entry[1]);
+            } else {
+                this.map.set(entry.IFluidDataStoreFactory.type, entry);
+            }
+        }
+    }
+
+    public async get(name: string): Promise<FluidDataStoreRegistryEntry | undefined> {
+        return this.map.get(name);
+    }
+}
+
+export class MultipleDataStoreRegistries implements IFluidDataStoreRegistry {
+    private readonly registries: IProvideFluidDataStoreRegistry[];
+
+    get IFluidDataStoreRegistry() { return this; }
+
+    constructor(...registries: IProvideFluidDataStoreRegistry[]) {
+        this.registries = registries;
+    }
+    public async get(name: string): Promise<FluidDataStoreRegistryEntry | undefined> {
+        for (const reg of this.registries) {
+            const res = await reg.IFluidDataStoreRegistry.get(name);
+            if (res !== undefined) {
+                return res;
+            }
+        }
+        return undefined;
```

Comment:
this is really worrying to me, as a registry may provide a default entry for unknown types, this can also result in an extern call for something like a remote catalog.

Reply:
Registry supports nesting. There is really no special protection by attempting to limit first layer and make it fixed - it's trivial to overcome. More in bigger comment below (I'll publish in a second).